Add Somfy MyLink support for Covers

bendews opened the pull request:
## Description:
This adds integration for the Somfy MyLink platform based on the Synergy JsonRPC API. It will automatically gather configured blinds and scenes from the MyLink hub and also allows additional options configured on a per-cover basis if required. For the time being, this is limited to reversal of the up/down method.

Thanks!

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#9063

## Example entry for `configuration.yaml` (if applicable):
```yaml
# This is the 'basic' configuration
somfy_mylink:
  host: 10.1.1.100
  system_id: mylink_id

# This is the 'advanced' configuration
somfy_mylink:
  host: 10.1.1.100
  system_id: mylink_id
  default_reverse: true
  entity_config:
    cover.outdoor_awning:
        reverse: false
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

bendews committed:
Added MyLink component

bendews committed:
Updated requirements.txt

bendews committed:
Fix lint issues

andrewsayre reviewed (changes requested):
Thanks for your contribution!  Please limit new components to a single platform. Add additional platforms on future PRs once we get the initial component in.

bendews commented:
@andrewsayre is that a new requirement? Haven't seen that before 😄 Just so I'm understanding correctly, I need to remove either Scene or Cover functionality..?

andrewsayre commented:
Yea, start with whichever one you think is more valuable. https://developers.home-assistant.io/docs/en/creating_component_code_review.html#5-limit-platforms-included-in-initial-pull-request

bendews committed:
Removed ‘Scene’ functionality

bendews commented:
@andrewsayre Oops, can't believe I missed that, thanks! 😄 I've removed the Scene component.

bendews committed:
Removed state restoration, as state is no longer tracked

awarecan commented:
It is harsh time for new integration as we are doing refactoring, we are now requesting all components has to have a [manifest file](https://github.com/home-assistant/architecture/issues/179).

Please rebase you branch against current `dev` branch, and create a `manifest.json` file under your component's folder.

You may need to do more rebase in the next few days as we are still in the progress of refactoring.

MartinHjelmare commented on homeassistant/components/somfy_mylink/__init__.py:
```diff
@@ -0,0 +1,74 @@
+Platform for the Somfy MyLink device supporting the Synergy JsonRPC API.
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/somfy_mylink/
```
We don't add the url to the docstring anymore. Just keep the very first line.

MartinHjelmare commented on homeassistant/components/somfy_mylink/__init__.py:
```diff
@@ -0,0 +1,74 @@
+    host = config[DOMAIN][CONF_HOST]
+    port = config[DOMAIN][CONF_PORT]
+    system_id = config[DOMAIN][CONF_SYSTEM_ID]
+    entity_config = config[DOMAIN].get(CONF_ENTITY_CONFIG, {})
```
We don't need the default dict here since we have it already in the config schema.

MartinHjelmare commented on homeassistant/components/somfy_mylink/__init__.py:
```diff
@@ -0,0 +1,74 @@
+    port = config[DOMAIN][CONF_PORT]
+    system_id = config[DOMAIN][CONF_SYSTEM_ID]
+    entity_config = config[DOMAIN].get(CONF_ENTITY_CONFIG, {})
+    if config[DOMAIN].get(CONF_DEFAULT_REVERSE):
```
There's a default for this key, so we can use `dict[key]`.

MartinHjelmare commented on homeassistant/components/somfy_mylink/__init__.py:
```diff
@@ -0,0 +1,74 @@
+        entity_config[CONF_DEFAULT_REVERSE] = True
+    try:
+        somfy_mylink = SomfyMyLinkSynergy(system_id, host, port)
+    except TimeoutError:
```
Don't we need to connect to possibly get a timeout error? We don't seem to connect when we instantiate `SomfyMyLinkSynergy`.

MartinHjelmare commented on homeassistant/components/somfy_mylink/cover.py:
```diff
@@ -0,0 +1,97 @@
+_LOGGER = logging.getLogger(__name__)
+DEPENDENCIES = ['somfy_mylink']
+DEFAULT_SUPPORTED_FEATURES = (SUPPORT_OPEN | SUPPORT_STOP | SUPPORT_CLOSE)
+ENTITY_ID_FORMAT = 'cover.{}'
```
Import this from the cover base component.

MartinHjelmare commented on homeassistant/components/somfy_mylink/cover.py:
```diff
@@ -0,0 +1,97 @@
+        self._device_class = device_class
+        self._closed = False
+        self._state = None
+        self._state_ts = None
```
We can remove these two state instance attributes. They aren't used.

MartinHjelmare commented on homeassistant/components/somfy_mylink/cover.py:
```diff
@@ -0,0 +1,97 @@
+        self._closed = False
+        self._state = None
+        self._state_ts = None
+        self._supported_features = DEFAULT_SUPPORTED_FEATURES
```
We don't need this it seems.

MartinHjelmare commented on homeassistant/components/somfy_mylink/cover.py:
```diff
@@ -0,0 +1,97 @@
+        self._name = name
+        self._reverse = reverse
+        self._device_class = device_class
+        self._closed = False
```
We can remove `self._closed`.

bendews committed:
Merge branch 'dev' into feature/somfy-mylink

bendews committed:
Add component manifest

bendews committed:
Remove documentation links in Docstring

bendews committed:
Removed redundant try/except block

bendews committed:
Removed default dict

bendews committed:
Removed features already implemented in default CoverDevice

bendews committed:
Removed attributes for tracking state

bendews committed:
Simplified loading of covers

No options exist other than reversal, so just check reversal status directly and update if needed

bendews committed:
Reimplemented is_closed property

bendews committed:
Import ENTITY_ID_FORMAT from base component

bendews committed:
Removed misc unused vars

bendews committed:
Update module docstrings to one line

bendews committed:
Removed too many blank lines, giving one back :)

bendews commented:
@awarecan & @MartinHjelmare, thanks for the feedback! I've fixed everything raised and also took the opportunity to simplify some of the logic - some of the complexity wasn't needed.

awarecan commented:
In order to satisfy our code coverage requirements, you need either

New files were added to .coveragerc.

Or

Tests have been added to verify that the new code works.

MartinHjelmare commented on homeassistant/components/somfy_mylink/cover.py:
```diff
@@ -0,0 +1,85 @@
+    except TimeoutError:
+        _LOGGER.error("Unable to connect to the Somfy MyLink device, "
+                      "please check your settings")
+        return False
```
Just `return`. Nothing is checking this return value.

bendews committed:
Return none on TimeoutError

bendews committed:
Added component to .coveragerc

bendews commented:
Thanks again @MartinHjelmare & @awarecan. I've skipped tests as this is assumed state, theres not anything meaningful to test from the HA side - I will likely add some tests to the underlying library in the future for parsing & validation of the returned data.

MartinHjelmare reviewed (approved):
Good!

MartinHjelmare commented on homeassistant/components/somfy_mylink/cover.py:
```diff
@@ -0,0 +1,85 @@
+    @property
+    def is_closed(self):
+        """Return if the cover is closed."""
+        pass
```
We should return `None` here.

Then set `assumed_state` property to `True`.

bendews replied:
@MartinHjelmare Oh nooo 😄 Apologies I didn't realise there was a specific property for assumed state! Should I raise a new PR?

MartinHjelmare replied:
Yes please!

bendews replied:
New PR #22922

MartinHjelmare reviewed (commented):
Sorry, I missed that we want this to entity to have assumed state.
